[API docs] More of it

glennsl opened the pull request:
Also fixed Js_null_udefined.bind and some tests

bobzhang commented on jscomp/others/js_null_undefined.ml:
```diff
@@ -33,11 +33,11 @@ external empty : 'a t = "undefined" [@@bs.val]
   match to_opt x with
-  | None ->  undefined
-  | Some x -> return (f  x [@bs])
+  | None -> Obj.magic x
```
can you write type annotations when using Obj.magic? otherwise it is unclear to me what you re doing here
```
(Obj.magic (x : 'a ) : 'b)
```
do you think it makes sense to provide a sugar like this `[%bs.unsafe_cat: 'a -> 'b ]`

glennsl replied:
What would that give us? Just mandatory type annotation? I'm down with that.

bobzhang commented:
Yeah, maybe `[%cast:'a -> 'b ]` ? since it is a general extension? 

reply@reply.github.com At: 03/16/17 09:53:05" data-digest="From: reply@reply.github.com At: 03/16/17 09:53:05" style="">
From: reply@reply.github.com At: 03/16/17 09:53:05
To: bucklescript@noreply.github.com
Cc: Hongbo Zhang (BLOOMBERG/ 731 LEX), comment@noreply.github.com
Subject: Re: [bloomberg/bucklescript] [API docs] More of it (#1380)


@glennsl commented on this pull request.   
In jscomp/others/js_null_undefined.ml: > @@ -33,11 +33,11 @@ external empty : 'a t = "undefined" [@@bs.val]
 
 let bind x f =
   match to_opt x with
-  | None ->  undefined
-  | Some x -> return (f  x [@bs])
+  | None -> Obj.magic x
 
What would that give us? Just mandatory type annotation? I'm down with that.  
—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or mute the thread.

glennsl commented:
I approve :)

glennsl commented:
@bobzhang I've annotated the types for Obj.magic, and added Js_float to the mix too. Beware that I've mucked around with `caml_*.ml` files too, dues to deprecations in `Js_float`

Let me know if you want me to add the generated html once you've given the go-ahead.

bobzhang commented on jscomp/others/js_null_undefined.ml:
```diff
@@ -33,11 +33,11 @@ external empty : 'a t = "undefined" [@@bs.val]
   match to_opt x with
-  | None ->  undefined
-  | Some x -> return (f  x [@bs])
+  | None -> (Obj.magic (x: 'a t): 'b t)
```
okay, this is more confusing, I understand why it works here, I will add type annotations to the parameter after merging : )

glennsl replied:
lol, yea sorry about that :)

bobzhang commented on lib/js/caml_format.js:
```diff
@@ -736,8 +737,8 @@ function caml_format_float(fmt, x) {
           var j = s.indexOf("e");
-          var exp = Number(s.slice(j + 1 | 0)) | 0;
-          if (exp < -4 || x$1 >= 1e21 || x$1.toFixed(0).length > prec$1) {
+          var exp = Caml_format.caml_float_of_string(s.slice(j + 1 | 0)) | 0;
```
this seems fishy?

bobzhang commented on jscomp/runtime/caml_format.ml:
```diff
@@ -526,13 +526,13 @@ let caml_format_float fmt x =
         let j = Bs_string.index_of !s "e" in 
-        let  exp = int_of_float @@ Js_float.of_any @@ Bs_string.slice_rest !s (j + 1)  in 
-        if exp < -4 || x >= 1e21 ||Bs_string.length (Js_float.toFixed x 0) > prec then 
+        let  exp = int_of_float @@ float_of_string @@ Bs_string.slice_rest !s (j + 1)  in 
```
I remember I used `Js_float.of_any` for some reasons

glennsl replied:
I could move the external in here and use that if you want. But if there's a subtle issue here, shouldn't there be a test for it that fails?

bobzhang replied:
`float_of_string` is ocaml semantics, like `float_of_string "1."`, any particular reason you want to make this change?

glennsl replied:
I want to get rid of `Js_float.of_any` because it's too polymorphic and error-prone,  but I have no issue with just moving it in here.

bobzhang replied:
can we rename it into `Js_float.ofString`

glennsl replied:
I suppose. `fromString` would be more consistent with the other `Js.*` APIs though.

bobzhang replied:
LGTM

glennsl commented:
@bobzhang I've added `Js_float.fromString` and replaced `float_of_string` in `caml_format.ml`. Also notice I added some functions to the `Js_int` module about two seconds before your last review.

bobzhang commented on jscomp/runtime/js_float.ml:
```diff
@@ -22,37 +22,290 @@
-[@@bs.send]
+@see <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NaN> MDN
+*)
+external nan : float = "NaN" [@@bs.val] 
```
to be consistent `_NaN`?

bobzhang commented on jscomp/test/mt.mli:
```diff
@@ -1,6 +1,8 @@
 type  eq =
   | Eq :  'a *'a  ->  eq
   | Neq : 'a * 'a ->  eq
+  | StrictEq :  'a *'a  ->  eq
```
in the future, when we add new ones, maybe just append in the end, to make diff less

bobzhang commented on lib/js/caml_format.js:
```diff
@@ -737,7 +737,7 @@ function caml_format_float(fmt, x) {
           var j = s.indexOf("e");
           var exp = Number(s.slice(j + 1 | 0)) | 0;
-          if (exp < -4 || x$1 >= 1e21 || x$1.toFixed(0).length > prec$1) {
+          if (exp < -4 || x$1 >= 1e21 || x$1.toFixed().length > prec$1) {
```
is this intentional?

bobzhang commented on jscomp/runtime/caml_format.ml:
```diff
@@ -526,13 +526,13 @@ let caml_format_float fmt x =
-        let  exp = int_of_float @@ Js_float.of_any @@ Bs_string.slice_rest !s (j + 1)  in 
-        if exp < -4 || x >= 1e21 ||Bs_string.length (Js_float.toFixed x 0) > prec then 
+        let  exp = int_of_float @@ Js_float.fromString @@ Bs_string.slice_rest !s (j + 1)  in 
+        if exp < -4 || x >= 1e21 ||Bs_string.length (Js_float.toFixed x) > prec then 
```
why `0` is ignored?

bobzhang replied:
I see, 0 is optional,

glennsl replied:
Yes, 0 is the default

bobzhang commented:
what do you think of `nan` vs `_NaN`?

glennsl commented:
`_NaN` would be more consistent with `Js.Math` and such. I'm for change.

glennsl commented:
Will you do it? If you do, remember all the references in doc comments

bobzhang commented:
Would you mind send a PR 😄
